### PR TITLE
Expand sorting of arity qualifiers to more attributes

### LINKED ISSUE #327: Expand `sort_function_exports` option to format (sort) attributes with pure lists of arity qualifiers
Currently, `default_formatter` formats this code, when `sort_function_exports` is set to `true`…

```erlang
%% assuming both the types and functions exist
-export_type([type_c/0, type_a/0, type_b/0, a_first_type/0]).
-export([b_fun/0, a_fun/0]).
-optional_callbacks([d_fun/0, c_fun/0]).
```

…like this…

```erlang
%% only the export attribute is sorted
-export_type([type_c/0, type_a/0, type_b/0, a_first_type/0]).
-export([a_fun/0, b_fun/0]).
-optional_callbacks([d_fun/0, c_fun/0]).
```

I would like it to be formatted like this…

```erlang
-export_type([a_first_type/0, type_a/0, type_b/0, type_c/0]).
-export([a_fun/0, b_fun/0]).
-optional_callbacks([c_fun/0, d_fun/0]).
```

It would be nice if the formatter were able to format (sort) other attributes than `-export` lists, as long as said attributes only contain arity qualifiers (e.g., `-dialyzer([{nowarn_function, a_fun/0}, no_return]).` should not be formatted (sorted)).

### PR DESCRIPTION
**Pull request by pablocostass**
Closes #327.

### CODE REVIEW

**Review by elbrujohalcon — CHANGES_REQUESTED** [review]
So, my goal was to _remove_ `export` as an exception… and make the code more generic.
If this is going to add 2 more exceptions (`export_type` and `optional_callback`), I'm not sure if we should go that way now.
Let's see if we can generalize this stuff so that our code is not forking on the tag of the attribute more than once.

File: src/formatters/default_formatter.erl [hunk 1]
```diff
@@ -1355,9 +1349,41 @@ lay_application(Name, Arguments, SpacesWithinParentheses, Ctxt) ->
...
+    when Tag =:= export_type; Tag =:= optional_callbacks ->
+    ArityQualifiers = erl_syntax:list_elements(OriginalAST),
+    case Ctxt#ctxt.sort_arity_qualifiers_match andalso Ctxt#ctxt.sort_arity_qualifiers of
+        true ->
+            SortFun = fun sort_arity_qualifiers_alphabetically/2,
+            SortedArityQualifiers = sort_arity_qualifiers(ArityQualifiers, Tag, SortFun),
+            erl_syntax:update_tree(OriginalAST, [SortedArityQualifiers]);
+        false ->
+            OriginalAST
+    end;
```
[the remark this answers is not in the document]
Comment (elbrujohalcon) [reply src/formatters/default_formatter.erl:1377]: To be clear: IF we can avoid the `is_list_skeleton` calls altogether… much better!

File: src/formatters/default_formatter.erl [hunk 2]
```diff
@@ -1355,24 +1349,55 @@ lay_application(Name, Arguments, SpacesWithinParentheses, Ctxt) ->
...
+        false ->
+            case erl_syntax:subtrees(Arguments0) of
+                [] ->
+                    %% node was a leaf node, skip
+                    [Arguments0];
+                [SubTrees0] ->
+                    SubTrees1 = lists:sort(fun sort_arity_qualifiers_alphabetically/2, SubTrees0),
+                    [erl_syntax:update_tree(Arguments0, [SubTrees1])]
+            end
+    end.
```
Comment (elbrujohalcon) [inline src/formatters/default_formatter.erl:1396]:
```suggestion
do_sort_arity_qualifiers(Arguments0) when is_list(Arguments0) ->
        [Sorted] = do_sort_arity_qualifiers(erl_syntax:list(Arguments0)),
        erl_syntax:list_elements(Sorted);
do_sort_arity_qualifiers(Arguments0) ->
        case erl_syntax:subtrees(Arguments0) of
            [] ->
                %% node was a leaf node, skip
                [Arguments0];
            [SubTrees0] ->
                SubTrees1 = lists:sort(fun sort_arity_qualifiers_alphabetically/2, SubTrees0),
                [erl_syntax:update_tree(Arguments0, [SubTrees1])]
        end.
```

**Review by elbrujohalcon — CHANGES_REQUESTED** [review]
Beautiful! We're almost there.
I just suggested a refactor.
And, as a bonus track (i.e. feel free to ignore this comment), you might want to try and see what happens if there are… 🙈 … macros… 😱 … there, like…

```erlang
-define(MY_FUN, a_very_long_function_name).
-export([a_completely_different_fun/0, ?MY_FUN/1, ?MY_FUN/2, ?MY_FUN/3, not_my_fun/4]).
```

File: src/formatters/default_formatter.erl (hunk 2 above)
Comment (pablocostass) [reply src/formatters/default_formatter.erl:1396]: Hehe, I very much approve of that refactor, which is how things used to be at first, before the comment about putting everything together in a single function clause; way more readable this way :D

**Comment by pablocostass** [comment]
Macros should behave as uppercase named functions. In any case, it's good to test them too, and I added your example to one of the test-cases :)

**Review by elbrujohalcon — APPROVED** [review]

**Pull request merged**